### ISSUE TITLE
wordPattern and tokenization are inconsistently applied

### ISSUE THREAD

**Issue by hoovercj** [Microsoft/vscode#37643]
We have noticed some inconsistent behavior when trying to leverage the `rename` functionality of a language extension with a word pattern that allows spaces. Depending on the context, the rename action would select the full "word" and other times it would only select a sub-word surrounded by spaces.

I have created a minimal [extension](https://github.com/hoovercj/textmate-bug-repro) that demonstrates the inconsistent behavior in features that rely on `getWordAtText` for word patterns that can include spaces.

The simplified extension relies on a very simple language definition with two key features:

1. A [word pattern](https://github.com/hoovercj/textmate-bug-repro/blob/master/language-configuration.json) that:
    * Allows spaces
    * Identifiers in this language are surrounded by double quoted strings and allows double quoted strings in the identifier by escaping them with an extra double quote. For example, `"An ""Identifier"" Example"`
2. A [simple grammar](https://github.com/hoovercj/textmate-bug-repro/blob/master/syntaxes/bug.tmLanguage.json):
    * A root token scope of "bug"
    * A "punctuation" token for semicolon

## Repro Steps

To observe the inconsistencies, launch this extension or install it from .vsix and open this extension directory.

Open the `./testfile.bug` with the following contents:

```
hi ; test ; "this "" is "" a test"

"this "" is "" a test"

hi ; test ; "this "" is "" a test"
```

Place the cursor in the word `is` on each of the lines and press `F2` to trigger a rename action. With the default "Dark+" theme, the behavior is as follows:

* Line 1: `is` appears in the rename box
* Line 3: `"this "" is "" a test"` appears in the rename box
* Line 5: `is` appears in the rename box

Line 3 behaves as expected, but lines 1 and 5 do not.

Now open the workspace settings and uncomment out the `editor.tokenColorCustomizations` settings.

Go back to `./testfile.bug` and repeat the rename steps from above. This time the behavior is as follows:

* Line 1: `"this "" is "" a test"` appears in the rename box
* Line 3: `"this "" is "" a test"` appears in the rename box
* Line 5: `is` appears in the rename box

Now Line 1 and line 3 behave as expected, but Line 5 still does not.

This is explained in a couple ways:

* Line 3 always works because there are no other tokens on the line, so the line can be treated as a single token and processed by the word pattern
* Line 1 works after enabling the coloring because coloring `;` helps [LineToken.tokenAt](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/core/lineTokens.ts#L141) return the correct token. With the `;` colorized it returns a token starting after the `;`, otherwise it returns a token starting at the beginning of the line. This then applies the word pattern to the text `hi ; test ; "this "" is "" a test"`. The word pattern matches on `hi` at this line in [wordHelper](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/model/wordHelper.ts#L121), and since it doesn't have a space in it then it tries to use `getWordAtPosFast` which won't work because the regex DOES allow spaces in it.
* Line 5 never works because [this line in textModelWithTokens](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/model/textModelWithTokens.ts#L437) claims it is not tokenized (an off by 1 error?). Due to this, it behaves the same as line 1 before adding the color because it treats the entire line as a token/starts searching from offset 0. If I debug, though, I can see that `this._lines` has 5 entries, and the 5th entry has a valid `_lineTokens` value, so it should not follow this path.

Now, to make it work in all three, simply add a blank line to the end of the file, forcing VS Code to recognize that line 5 is tokenized. If you go back to `./testfile.bug` and repeat the rename steps from above the behavior is as follows:

* Line 1: `"this "" is "" a test"` appears in the rename box
* Line 3: `"this "" is "" a test"` appears in the rename box
* Line 5: `"this "" is "" a test"` appears in the rename box

**Comment by alexandrudima** [comment]
`IModel.getWordAtPosition` tries to first find the token (or the two tokens) that contain the position and then proceeds to look for a word within that token.

This is tricky and perhaps doesn't make a lot of sense, as tokens that are equal in color (based on theme) get merged since about 1 year.

We can remove some of that code and have `getWordAtPosition` ignore tokens, except in the mixed language case.

**Comment by alexandrudima** [comment]
I have fixed the off-by-one issue and I have modified that tokens boundaries are not considered anymore, except in the case where there are mixed languages.

The original repro gets reduced:
* color customizations have no effect
* line 1 behaves the same as line 5

There is still a discrepancy between lines 1, 3, 5. It boils down to the call to `getWordAtText` which now gets the entire line passed in all the time. For lines 1 and 5, it finds `hi` as the first word on the line and concludes that it can go down the fast route. For line 3, it finds the entire line as a match and proceeds down the slow route.
